File: dominoAndTrominoTiling.py
# problem: https://leetcode.com/problems/domino-and-tromino-tiling/
# Runtime: 64 ms, faster than 15.13% of Python3 online submissions for Domino and Tromino Tiling.
# Memory Usage: 16.9 MB, less than 8.46% of Python3 online submissions for Domino and Tromino Tiling.

class Solution:
    def numTilings(self, n: int) -> int:
        cache = {}
        def backtrack(first, second):
            if (first, second) in cache:
                return cache[(first, second)]
            
            if first > n or second > n:
                return 0
            
            if first == second == n:
                return 1
            
            total = 0
            if first == second:
                total += backtrack(first + 1, second + 1)
                total += backtrack(first + 2, second + 2)
                total += backtrack(first + 1, second + 2)
                total += backtrack(first + 2, second + 1)
            elif first > second:
                total += backtrack(first + 1, second + 2)
                total += backtrack(first, second + 2)
            else:
                total += backtrack(first + 2, second + 1)
                total += backtrack(first + 2, second)
        
            cache[(first, second)] = total
            return total % (10 **9 + 7)
        
        return backtrack(0, 0)

File: dominoAndTrominoTiling.cpp
// problem: https://leetcode.com/problems/domino-and-tromino-tiling/description/
// Runtime
// 23 ms
// Beats
// 8.51%
// Memory
// 48.3 MB
// Beats
// 5.15%

#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
  int numTilings(int n) {
    vector<vector<int>> dp(n + 1, vector<int>(n + 1, 0));
    dp[n][n] = 1;
    int mod = 1e9 + 7;
    for (int i = n; i >= 0; --i) {
      for (int j = n; j >= 0; --j) {
        if (i == j) {
          if (i > 1) {
            dp[i - 2][j - 2] += dp[i][j];
            dp[i - 2][j - 2] %= mod;
            dp[i - 1][j - 2] += dp[i][j];
            dp[i - 1][j - 2] %= mod;
            dp[i - 2][j - 1] += dp[i][j];
            dp[i - 2][j - 1] %= mod;
          }
          if (i > 0) {
            dp[i - 1][j - 1] += dp[i][j];
            dp[i - 1][j - 1] %= mod;
          }
        } else if (i + 1 == j) {
          if (j > 1) {
            dp[i - 1][j - 2] += dp[i][j];
            dp[i - 1][j - 2] %= mod;
            dp[i][j - 2] += dp[i][j];
            dp[i][j - 2] %= mod;
          }

        } else if (j + 1 == i) {
          if (i > 1) {
            dp[i - 2][j - 1] += dp[i][j];
            dp[i - 2][j - 1] %= mod;

            dp[i - 2][j] += dp[i][j];
            dp[i - 2][j] %= mod;
          }
        }
      }
    }
    return dp[0][0];
  }
};
